core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -600,6 +600,11 @@ class UnifiedLog(@volatile var logStartOffset: Long,
    */
   def hasOngoingTransaction(producerId: Long, producerEpoch: Short): Boolean = lock synchronized {
     val entry = producerStateManager.activeProducers.get(producerId)
+    // With transactions V2, if we see a future epoch, we are likely in the process of completing the previous transaction.
```

Comment:
To confirm, we shouldn't expect a future epoch, right?

Reply:
Hmmm. We can see a future epoch in the case where the next transaction has started while the previous one is completing. For V2 this can happen due to the reordering of handling the produce and add partitions

Reply:
Ah right. This in theory could be handled still, but if the transaction completes while adding the partition we hit the race condition. As the code is now, we don't want to handle this case.